### PR TITLE
Add a Sled datastore

### LINKED ISSUE #79: Sled backend?
Hello! Excellent work on this project so far. I'm stoked to try it out. Any plans to implement a Sled backend for this project per chance?

Either way, it looks like the interface is pretty straightforward, and Sled does offer a transactional interface. Thoughts on implementation?

### PR DESCRIPTION
**Pull request by jespersm**
Fixes #79.

### CODE REVIEW

**Comment by ysimonson** [comment]
Thanks for bringing this together @jespersm! It's a big PR, so it'll take me a little while to go over everything -- but I'm excited for us to have sled support, especially with the performance results you posted.

**Comment by jespersm** [comment]
> Thanks for bringing this together @jespersm! It's a big PR, so it'll take me a little while to go over everything -- but I'm excited for us to have sled support, especially with the performance results you posted.

Performance numbers are on macOS, by the way.

While microbenchmarking is fine, a proper performance test would examine parallel performance on contended keys.

**Comment by ysimonson** [comment]
Yup, I'd love to have some more fleshed out benchmarks, but it's all we have at the moment. Results will likely vary by workload type, but at least sled now performs roughly the same in the microbenchmarks. Running against the wikipedia dataset may give a better idea.

File: lib/src/sledds/datastore.rs [hunk 1]
```diff
@@ -0,0 +1,494 @@
...
+fn remove_nones_from_iterator<I, T>(iter: I) -> impl Iterator<Item = Result<T>>
+where
+    I: Iterator<Item = Result<Option<T>>>,
+{
+    iter.filter_map(|item| match item {
+        Err(err) => Some(Err(err)),
+        Ok(Some(value)) => Some(Ok(value)),
+        _ => None,
+    })
+}
```
Comment (ysimonson) [inline lib/src/sledds/datastore.rs]: Could you move this method to `lib/src/util.rs`, and reuse it both here and in the rdb implementation?

**Comment by ysimonson** [comment]
Are you familiar with the frontend (i.e., `bin`?) It'd be good to wire in support for this into it. If you're not familiar with it, though, I can tackle that.

File: lib/src/sledds/datastore.rs (hunk 1 above)
Comment (jespersm) [reply lib/src/sledds/datastore.rs]: Yep, I can do that

**Comment by vn971** [comment]
Any updates on the merge request? Seems like some small-ish questions are there, otherwise we could merge it?

**Comment by ysimonson** [comment]
@jespersm took it most of the way there. @vn971 if you or someone else is willing to do the last mile of work, I'm happy to review.

**Comment by jespersm** [comment]
I'll take a look at adding support in 'bin'

**Comment by jespersm** [comment]
The tests for 'bin' seem to stall whenever trying to set up the client datastore, even for master branch, both benches and tests.

**Comment by ysimonson** [comment]
I see, will take a look shortly

**Comment by ysimonson** [comment]
I just merged in [a few touchups to the CI infrastructure.](https://github.com/indradb/indradb/pull/86) Part of the goal there was to try to repro the issue on this PR, but I didn't manage to. But the PR also reorders the tests, such that lib tests run before bin.

Could you merge or rebase master into this branch? It won't fix the issue, but I'm hoping that running the lib tests first may shine some light on what's going on.

**Comment by jespersm** [comment]
I'm closing this in favor of #89.

**Pull request closed without merging**